File: Data Structures & Algorithms/minimum-interval-including-query/submission-0.py
from typing import List
import heapq

class Solution:
    def minInterval(self, intervals: List[List[int]], queries: List[int]) -> List[int]:
        # Sort intervals by left endpoint
        intervals.sort(key=lambda x: x[0])
        n = len(intervals)

        # Pair each query with its original index, then sort by query value
        indexed_queries = sorted([(q, i) for i, q in enumerate(queries)])

        ans = [-1] * len(queries)
        heap = []  # min-heap of (length, right)
        i = 0      # pointer over intervals

        for q, idx in indexed_queries:
            # Add all intervals that start on/before q
            while i < n and intervals[i][0] <= q:
                L, R = intervals[i]
                heapq.heappush(heap, (R - L + 1, R))
                i += 1

            # Remove intervals that end before q (can't cover q)
            while heap and heap[0][1] < q:
                heapq.heappop(heap)

            # Top of heap (if any) is the shortest interval covering q
            if heap:
                ans[idx] = heap[0][0]  # the length

        return ans

File: Data Structures & Algorithms/valid-palindrome-ii/submission-0.java
class Solution {
    public boolean validPalindrome(String s) {
        char[] ch = s.toCharArray();
        int left = 0;
        int right = s.length()-1;
        while(left<right){
            if(ch[left]!=ch[right]){return checkPalindrome(s,left+1,right)||checkPalindrome(s,left,right-1);}
            else{left++;right--;}
        }
        return true;
    }

    private boolean checkPalindrome(String s, int left, int right){
        char[] ch = s.toCharArray();
        while(left<right){
            if(ch[left]!=ch[right])return false;
            left++;right--;
        }
        return true;
    }
}